### PR TITLE
Tentative simplification of the unfocusing function

### PR DESCRIPTION
**Pull request by olaure01**

### CODE REVIEW

**Comment by etiennecallies** [comment]
Ouch conflicts :confounded: I will take your ideas and apply them on my coed, there are good ones!

**Comment by etiennecallies** [comment]
I tried to use theta as set like you did but, the order is not guaranteed when calling `elements`.
```
module Set_test = Set.Make(struct type t = string let compare = compare end);;
let a = Set_test.empty;;
let a = Set_test.add "hello" a;;
let a = Set_test.add "world" a;;
let a = Set_test.add "you" a;;
let a = Set_test.add "rock" a;;
Set_test.elements a;;
```
will print
```
- : Set_test.elt list = ["hello"; "rock"; "world"; "you"]
```

**Comment by olaure01** [comment]
It is guaranteed to be ordered with respect to `compare`:
https://ocaml.org/api/Set.S.html#VALelements
This should be enough here.

File: focused_proof.ml [hunk 1]
```diff
@@ -402,77 +397,77 @@ let rec unfocalize_proof theta = function
...
-                    let n_theta = List.length theta in
-                    let n_gamma = List.length gamma in
-                    let permutation = List.init n_theta (fun k -> 1 + k)
-                        @ List.init n_gamma (fun k -> 1 + n_theta + k)
-                        @ [0]
-                        @ List.init (List.length tail) (fun k -> 1 + n_theta + n_gamma + k) in
-                    Exchange_proof (new_sequent, permutation, premise)
+                | Async (theta, gamma, (Whynot e) :: tail) ->
+                    if Set_formula.mem e theta then
+                      Weakening_proof (map_wn (Set_formula.elements theta) @ gamma, e, tail, unfocus_proof (List.hd focused_premises))
```
Comment (etiennecallies) [inline focused_proof.ml:424]: What does it bring? This case will be handled by the "weakening simplification" at the end, won't it?

Comment (etiennecallies) [reply focused_proof.ml:424]: understood, it is to avoid doublon in theta.

Comment (olaure01) [reply focused_proof.ml:424]: It `theta` is not given as an argument anymore, this allows to preserve the invariant that `unfocus_proof` is a proof with conclusion `map_wn (Set_formula.elements theta) @ gamma @ lambda` if the focused proofs has conclusion `Async (theta, gamma, lambda)`.

File: focused_proof.ml [hunk 2]
```diff
@@ -402,77 +397,77 @@ let rec unfocalize_proof theta = function
...
+                    else 
+                      let premise = unfocus_proof (List.hd focused_premises) in
+                      let theta_head, theta_tail = head_tail e (Set_formula.elements theta) in
```
Comment (etiennecallies) [inline focused_proof.ml:427]: should be theta of premise, shouldn't it?

Comment (olaure01) [reply focused_proof.ml:427]: Yep!

**Comment by etiennecallies** [comment]
> It is guaranteed to be ordered with respect to `compare`:
> https://ocaml.org/api/Set.S.html#VALelements
> This should be enough here.

Indeed, that's great simplification, you can have a look at my PR #77 now!

**Comment by olaure01** [comment]
@etiennecallies is there still some code to be used from here, or should we close this PR?

**Comment by etiennecallies** [comment]
Nope. Thanks for you work! Everything has been done on #77.

**Pull request closed without merging**